lte/gateway/c/session_manager/SetMessageManagerHandler.cpp

```diff
@@ -72,50 +72,95 @@ void SetMessageManagerHandler::SetAmfSessionContext(
+  if (cfg.common_context.sm_session_state() == RELEASED_4) {
+    if (cfg.common_context.sm_session_version() != 0) {
+      /* This is a serious sync error, as AMF passing wrong version values
+       * for this session to release. As session version can never be 0.
+       * Print the error message and keep releasing the respective session
+       */
+      MLOG(MERROR) << "Wrong version received from AMF for IMSI " << imsi
+                   << " but continuing release request";
+    }
+    MLOG(MINFO) << "Release request for session from IMSI: " << imsi << " DNN "
+                << dnn;
+    // Requested message from AMF to release the session
+    m5g_enforcer_->get_event_base().runInEventBaseThread(
+        [this, cfg, imsi, dnn]() {
           auto session_map = session_store_.read_sessions({imsi});
-          send_create_session(session_map, imsi, session_id, cfg);
-        }
-      });
+          initiate_release_session(session_map, dnn, imsi);
+        });
+  }
+  // Else the requested message is for establishment or modification
+  else {
+    // The Event Based main_thread invocation and runs to handle session state
+    m5g_enforcer_->get_event_base().runInEventBaseThread([this, cfg, imsi,
+                                                          dnn]() {
+      std::string session_id = id_gen_.gen_session_id(imsi);
+      MLOG(MINFO) << "Requested session from UE with IMSI: " << imsi
+                  << " Generated session context ID " << session_id;
+
+      /* Message may be intial or modification message. Only taken care
+       * intial message. Check if it's initial message
+       */
+      if ((cfg.rat_specific_context.m5gsm_session_context().rquest_type() ==
+           INITIAL_REQUEST) &&
+          (cfg.common_context.sm_session_state() == CREATING_0)) {
+        /* it's new UE establisment request and need to create the session
+         * context
+         */
+        MLOG(MINFO)
+            << "AMF request type INITIAL_REQUEST and session state CREATING";
+        /* Read the SessionMap from global session_store,
+         * if it is not found, it will be added w.r.t imsi
+         */
+        auto session_map = session_store_.read_sessions({imsi});
+        send_create_session(session_map, imsi, session_id, cfg, dnn);
+      }
+    });
+  }  // end of else
 }
 
 /* Creeate respective SessionState and context*/
 void SetMessageManagerHandler::send_create_session(
     SessionMap& session_map, const std::string& imsi,
-    const std::string& session_id, const SessionConfig& cfg) {
+    const std::string& session_id, const SessionConfig& cfg,
+    const std::string& dnn) {
+  /* If it is new session to be created, check for same DNN exists
+   * for same IMSI, i.e if IMSI found and respective DNN found in
+   * SessionStore, then return from here and nothing to do
+   * as already same session exist, its duplicate request
+   */
+  auto imsi_exist = session_map.find(imsi);

```

Comment:
Couldn't we use the search function here as you do in other places?

Reply:
If its duplicate APN we need to log as duplicate, that reason we got the session_map first, then checking the dnn.

Reply:
Can't you just reverse the condition?
`if` will be fale in case no session or no apn found
```
    SessionSearchCriteria criteria(imsi, IMSI_AND_APN, dnn);
    auto session_it = session_store_.find_session(session_map, criteria);
--> if (session_it) {
        MLOG(MERROR) << "Duplicate request of same DNN " << dnn << " of IMSI "
                     << imsi << " nothing to do";
        return;
    }
	auto session_map_ptr = std::make_shared<SessionMap>(std::move(session_map));
```